axiom/nr_segment_message.c

```diff
@@ -206,12 +208,15 @@ bool nr_segment_message_end(nr_segment_t** segment_ptr,
    * Additionally, because it makes http calls under the hood,
    * we don't want additional external calls created for this same txn.
    * Therefore, we delete all children of the message segment.
+   * By destroying the tree we are able to destroy all descendants vs just
+   * destroying the child which then reparents all it's children to the segment.
    */
   if (segment) {
     for (size_t i = 0; i < nr_segment_children_size(&segment->children); i++) {
-      nr_segment_t* child = nr_segment_children_get(&segment->children, i);
-      nr_segment_discard(&child);
+      child = nr_segment_children_get(&segment->children, i);
+      nr_segment_destroy_tree(child);
```

Comment:
The header docs for this function say " * WARNING : This should only be called during transaction destruction."  Do we know this is safe at this point?

Reply:
Yes, the warning is because it permanently walks the tree and destroys all children and their children.  In this case, where we always do not want message segments to have children, that is exactly what we want.  In most other cases, discard is more appropriate since it reparents a discarded child segment's children to the discarded child segment's parent.